### PR TITLE
fix: avoid complex SQL when incrementing streaks

### PR DESCRIPTION
**Pull request by DragosIorgulescu**
- make use of the view record we just created, or the last view record if we did not create one when checking if we need to increment streaks
- avoid timezone adjustments when updating `lastViewAt` -- we simply store the same date which is present on the `view.timestamp`, so when we check against it, we know we are checking against the same timezone and timestamp essentially

ref MI-313

### CODE REVIEW

**Review by idoshamun — COMMENTED** [review]
Since we now migrated all users, should we consider this comment? https://github.com/dailydotdev/daily-api/pull/1836/files#diff-9ce2515f0e6544d6bcb887eed3d4a732981b644468f04b7f3985c000458199e7R43

Also what about the migration script, do we need to fix it?

File: src/workers/newView.ts
Comment (capJavert) [inline src/workers/newView.ts:29]: @kopancek can you take a look at this changes since you were the author of original query. I am all up for removing it since typeorm logic is much easier to parse mentally but I am not 100% if there is some edge case this new implementation does not cover. 

I see some tests fail so something might be off.

Comment (DragosIorgulescu) [reply src/workers/newView.ts:29]: looking into specs now, they may be failing due to how they are set up, but yes, would love @kopancek eyes on this as well 🙏

**Comment by DragosIorgulescu** [comment]
> Since we now migrated all users, should we consider this comment?

@idoshamun thought about that, but did not want to include that in the same PR, I'll push a separate one for this cleanup 🙏

**Comment by DragosIorgulescu** [comment]
updated specs to explicitly validate we are setting `lastViewAt` correctly, and the streaks seed to reflect the fact that we are no longer storing the `lastViewAt` in the user's timezone 

_this does not require a re-run of the seed, because it was correctly converting the `view.timestamp` to the same timezone as the `lastViewAt` value, so it was correctly comparing apples to apples -- this was not happening in the `newView` worker_

**Review by idoshamun — DISMISSED** [review]
Please wait for @capJavert approval please

File: __tests__/workers/newView.ts
Comment (kopancek) [inline __tests__/workers/newView.ts:179]: This change does not seem correct to me. Previously we were comparing the lastViewAt on the `user_streak` table, now we do it against the `view` table. 

If we want to keep the same test behavior, we should take the `lastViewAt` from the `streak` object you create above.

File: __tests__/workers/newView.ts
Comment (kopancek) [inline __tests__/workers/newView.ts:330]: `defaultStreak` does not contain the `lastViewAt` timestamp. That's why the test was sending it before. Sending just `defaultStreak` as the last parameter here means, that we do not even try to compare the dates anymore. Which again, seems to defeat the purpose of the test.

File: __tests__/workers/newView.ts
Comment (kopancek) [inline __tests__/workers/newView.ts:336]: Same as above

File: src/workers/newView.ts
Comment (kopancek) [inline src/workers/newView.ts:28]: I think if we extract the `isNewView` into the upper scope, the code would look nicer. E.g.:
```typescript
let view = existing;
const isNewView = !existing ||
    entity.timestamp.getTime() - existing.timestamp.getTime() > ONE_WEEK
if (isNewView) {
    view = await repo.save(entity)
}
return { view, isNewView };
```

File: src/workers/newView.ts
Comment (kopancek) [inline src/workers/newView.ts:43]: Since we are touching the file, we can basically get rid of this logic now. The migration has been run, so all users will have a record in the `user_streak` table AFAIK.

File: src/workers/newView.ts
Comment (kopancek) [inline src/workers/newView.ts:146]: Why do we need the `!latestView` condition as well? If `didSave` is `true`, `latestView` will always be defined.

If you want, we can instead base the logic on the `!latestView`. But then we can get rid of the `didSave` variable completely and rely on `latestView` being non null instead.

File: src/workers/newView.ts
Comment (kopancek) [inline src/workers/newView.ts:69]: Both `lastViewAt` and `timestamp` are in UTC by default. This code does not deal with user timezones at all and will increment the reading streak only with UTC in mind, not the user timezone. I'm not sure that is the intention.

Imagine the following scenarios:
# User in PDT
1. User is in PDT (UTC -08:00)
2. User reads an article at 12:00:00 local time, which is 20:00 UTC. We increment the streak
3. User reads article at 19:00:00 local time, same date, which is 3AM UTC the next day. We end up incrementing the streak, although we should not have.

# User in Sydney
1. User is in Sydney (UTC +10:00)
2. User reads an article at 16:00 local time, which is 2AM UTC. We increment the streak.
3. User reads an article the next day at 7:00 local time, which is 17:00 UTC, the same day as was the 2AM UTC view. We do not increment the streak, even if should have done it.

Does that make sense?

**Review by kopancek — CHANGES_REQUESTED** [review]
I'm not sure the current code takes user timezone in mind at all. The tests also seem to have been crippled by the changes and do not test if the `lastViewAt` column on the streak is what we expected. Was that the intention?

File: src/workers/newView.ts
Comment (DragosIorgulescu) [reply src/workers/newView.ts:43]: yep, already mentioned will be cleaning this up in a separate PR 🙏

File: __tests__/workers/newView.ts
Comment (DragosIorgulescu) [reply __tests__/workers/newView.ts:179]: we actually don't want to keep the same behaviour, the goal here is not to compare against the date coming in, but against the timestamp of the last view record. 

essentially what we are doing here is asserting that after the worker runs, our streak `lastViewAt` correctly matches the  `timestamp` of the last view -- we intentionally don't want to compare it against the date being passed in

File: __tests__/workers/newView.ts
Comment (DragosIorgulescu) [reply __tests__/workers/newView.ts:330]: yep, and it's okay that it does not contain it, as we are no longer comparing against the incoming date (which would require us to do timezone conversion shenanigans) 🙏

File: src/workers/newView.ts
Comment (DragosIorgulescu) [reply src/workers/newView.ts:28]: sure thing, good point 👍

**Comment by idoshamun** [comment]
> It does not matter here what the user timezone is, as when the View is created, the timestamp is stored at the current time converted to the db timezone, which is UTC. And as long as the streaklastViewAt is in the same timezone, we are doing a valid comparison.

This assumption is wrong isn't it? Because you want to check the two timestamps are on a different day. Day is defined by the time zone. Milan provides good examples that break this assumption

**Comment by DragosIorgulescu** [comment]
My logic was more on making sure we don't have a mismatch between views and streaks, if we store views in UTC, do we end up displaying views in the right timezone?

**Comment by idoshamun** [comment]
@DragosIorgulescu yes we always convert them. User time zone is dynamic so there's no point in storing the timestamp in the right timezones

**Comment by kopancek** [comment]
> My logic was more on making sure we don't have a mismatch between views and streaks, if we store views in UTC, do we end up displaying views in the right timezone?

It depends on how the logic to show the views is implemented. You can do the same shenanigan, e.g. `AT TIMEZONE` of the user, something to check I guess.

The DB always stores all times in UTC. The NodeJS code on the backend uses UTC by default as well. This is not an issue by itself, until you start comparing the dates from the timestamp, which we need to do for streaks.

We can still cut away the complex SQL logic and do the DB queries with ORM, but we need to then do timezone conversion on the backend. E.g. using `moment` timezone functionality. This should be fine to add as a new dependency if we don't already have it.

File: __tests__/workers/newView.ts
Comment (kopancek) [reply __tests__/workers/newView.ts:330]: If we do not send the `lastViewAt`, then the `expectedStreak` object does not contain the `lastViewAt` key and we do not do this comparison at all:
```typescript
if (key === 'lastViewAt') {
        expect(streak?.[key]).toEqual(lastView.timestamp);
      } else {
        expect(streak?.[key]).toEqual(expectedStreak[key]);
      }
```

**Comment by DragosIorgulescu** [comment]
ok, makes sense, nevermind these changes then 😅  I'll close this PR out without merging. Thanks for the input @kopancek , a very good sanity check on my ramblings 🙏

**Pull request closed without merging**
